### PR TITLE
tm: new weight-based call-termination distribution

### PR DESCRIPTION
**Pull request by marcocapetta**
#### Pre-Submission Checklist
- [x] Commit message has the format required by CONTRIBUTING guide
- [x] Commits are split per component (core, individual modules, libs, utils, ...)
- [x] Each component has a single commit (if not, squash them into one commit)
- [x] No commits to README files for modules (changes must be done to docbook files
in `doc/` subfolder, the README file is autogenerated)

#### Type Of Change
- [ ] Small bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)

#### Checklist:
- [ ] PR should be backported to stable branches
- [x] Tested changes locally
- [x] Related to issue #2167

#### Description
Extended t_load_contacts function with the parameter 'mode'.
It can be used to define the algorithm to use for ordering the contacts.
If 0, or missing, the standard algorithm with be used.
If 1, the new weight-based algorithm will be used.

The new algorithm uses the q value as the weight of the branch.
Higher is the q value, higher is the probability to be tried as first. Equals q values means equals probability to be tried.
Branches with q values equals to 0 or lower are not considered by the ordering algorithm, but just added at the end of the list as backup option if all other branches fail.

### CODE REVIEW

**Comment by miconda** [comment]
I looked quickly trying to see how the existing code was changed, which seems to be mainly moved in a dedicated function. I didn't dig in the new algorithm (mode), being specific for the feature wanted with this PR.

Maybe @juha-h wants to check a bit and see if the old mode (now STANDARD) is not affected somehow.

Otherwise, if nobody else has comments, I am fine to merge.

File: src/modules/tm/t_serial.c [hunk 1]
```diff
@@ -159,26 +164,171 @@ void add_contacts_avp(str *uri, str *dst_uri, str *path, str *sock_str,
...
+ * only one contact at a time has to ring.
+ */
+int t_load_contacts_proportional(struct contact *contacts, char *sock_buf, int n, unsigned short q_total)
+{
+	int q_remove, n_rand, idx;
+	struct contact *curr;
+	sr_xavp_t *lxavp = NULL;
+
+	/* Initialize the random generator */
+	kam_srand(time(NULL));
```
Comment (henningw) [inline src/modules/tm/t_serial.c]: This is not necessary (and will also overwrite the state for the global one). The random generator is initialized during startup.

File: src/modules/tm/t_serial.h [hunk 2]
```diff
@@ -28,14 +28,19 @@
 #ifndef _T_SERIAL_H_
 #define _T_SERIAL_H_
 
+#define STANDARD 0
```
Comment (henningw) [inline src/modules/tm/t_serial.h]: Makes probably sense to qualify the defines, e.g. with T_LOAD_ prefix or similar. Also consider to move them into the respective .c file, if they are not necessary in the .h

**Comment by henningw** [comment]
Thank you for the change. @juha-h - comments from your side as well?

**Comment by henningw** [comment]
Merged - eventual further changes could be done directly in git master.

**Pull request merged**